Configurable HorizontalPodAutoscaler

gliush opened the pull request:
I introduce an algorithm-agnostic HPA object configuration that will configure each particular HPA scaling behavior. 

KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

This PR contains only API changes for now. The business logic will be introduced in a separate PR. For now I keep all the changes in my local repo until the current PR is approved:
https://github.com/gliush/kubernetes/pull/2


**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Different applications may have different business values, different logic and may require different scaling behaviors. 
At the moment, there’s only one cluster-level configuration parameter that influence how fast the cluster is scaled down:
[--horizontal-pod-autoscaler-downscale-stabilization-window](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details)   (default to 5 min)

And a couple of hard-coded constants that specify how fast the cluster can scale up:
- [scaleUpLimitFactor](https://github.com/kubernetes/kubernetes/blob/7ba79c31838632fa535f9aaab516276c2160a09d/pkg/controller/podautoscaler/horizontal.go#L55) = 2.0
- [scaleUpLimitMinimum](https://github.com/kubernetes/kubernetes/blob/7ba79c31838632fa535f9aaab516276c2160a09d/pkg/controller/podautoscaler/horizontal.go#L56) = 4.0

This PR introduces an algorithm-agnostic HPA object configuration that will configure each particular HPA scaling behavior.
For both directions (scale up and scale down) it will be possible to specify one or several `behavior` that will control how the HPA controller will scale the resources. Effectively each `behavior` is a specification of "how many percents/pods could be added/removed per some period of time".
Additionally, each direction might have a `StabilizationWindowSeconds` parameter set to gather recommendations for some time and pick the safest change.

For more information and motivation read the KEPs (links are above). 


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes  https://github.com/kubernetes/kubernetes/issues/39090
Fixes https://github.com/kubernetes/kubernetes/issues/65097
Fixes https://github.com/kubernetes/kubernetes/issues/69428

It covers partly  https://github.com/kubernetes/kubernetes/issues/56335


**Special notes for your reviewer**:
The API changes are backward compatible. All current defaults are kept the same.

**Does this PR introduce a user-facing change?**:
```release-note
Added the HPA API, that allows scale behavior to be configured through the HPA
`behavior` field. Behaviors are specified separately for scaling up and down. In
each direction a stabilization window can be specified as well as a list of
policies and how to select amongst them. Policies can limit the absolute number
of pods added or removed, or the percentage of pods added or removed.
```

gliush commented:
/sig autoscaling

gliush commented:
@thockin: Could you have a look? @mwielgus says that you're the most familiar with the matter.
I can't do it by myself, sorry.

thockin commented:
Is there a KEP for this?  Generally we want to see the design work independent of the code.

I am happy to take up the API review, but can't really do that until I know the domain experts (e.g. @mwielgus) are happy with the design.

thockin commented:
I see there's a doc but that's not a KEP

thockin commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -69,6 +69,9 @@ type HorizontalPodAutoscalerSpec struct {
 	Metrics []MetricSpec `json:"metrics,omitempty" protobuf:"bytes,4,rep,name=metrics"`
+	// constraints contain HPA object configuration parameters
+	// that configure each particular HPA scaling behaviour.
+	Constraints HorizontalPodAutoscalerScaleConstraints `json:"constraints,omitempty" protobuf:"bytes,5,opt,name=constraints"`
```
I am not sure that 'constraints' is the best word here.  It's OK, but I want to ask if we considered an y others

mwielgus replied:
That is the best we have :-|, but we will happily consider an alternative.

thockin commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -117,6 +120,26 @@ type MetricSpec struct {
+type HorizontalPodAutoscalerScaleConstraints struct {
+	// scaleUpPercent specifies the scale up relative speed, in percentages
+	// i.e. if scaleUpPercent = 150 , then we can add 150% more pods (10 -> 25 pods)
+	ScaleUpPercent *resource.Quantity `json:"scaleUpPercent,omitempty" protobuf:"bytes,1,opt,name=scaleUpPercent"`
```
why Quantity?  That doesn't seem right.  Do you really want to be able to specify "0.1" or "3Mi" ?  Why is "int32" not sufficient?

mwielgus replied:
The idea behind quantity is that all values are per minute. So if you want to add/remove at most 1 pod every 2 minutes you need to put 0.5 here. 

Quantity is used for consistency with pods. We can however use ints in percent-like values if you prefer.

mwielgus commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -117,6 +120,26 @@ type MetricSpec struct {
+	// i.e. if scaleDownPercent = 60 , then we can remove 60% pods (10 -> 4 pods)
+	ScaleDownPercent *resource.Quantity `json:"scaleDownPercent,omitempty" protobuf:"bytes,2,opt,name=scaleDownPercent"`
+	// scaleUpPods specifies the scale up absolute speed, in number of pods
+	// i.e. if scaleUpPods = 5 , then we can add 5 more pods (10 -> 15)
```
Please add comment about fractional values.

gliush replied:
Will do

gliush replied:
Done.

thockin commented:
Why not allow the user to specify the denominator?  "1 pod per 2 minutes"
is more expressive than "0.5 pods per minute".

On Fri, Mar 1, 2019 at 2:09 PM Marcin Wielgus <notifications@github.com>
wrote:

> *@mwielgus* commented on this pull request.
> ------------------------------
>
> In staging/src/k8s.io/api/autoscaling/v2beta2/types.go
> <https://github.com/kubernetes/kubernetes/pull/74525#discussion_r261774126>
> :
>
> > @@ -117,6 +120,26 @@ type MetricSpec struct {
>  // MetricSourceType indicates the type of metric.
>  type MetricSourceType string
>
> +// HorizontalPodAutoscalerScaleConstraints configures the scaling velocity
> +// by specifying the "absolute" value (in number of pods) and "relative" values (in percents)
> +// All the parameters of the struct are "per minute".
> +// For each scale direction (Up or Down) if both parameters are specified
> +// the largest constraint is used.
> +type HorizontalPodAutoscalerScaleConstraints struct {
> +	// scaleUpPercent specifies the scale up relative speed, in percentages
> +	// i.e. if scaleUpPercent = 150 , then we can add 150% more pods (10 -> 25 pods)
> +	ScaleUpPercent *resource.Quantity `json:"scaleUpPercent,omitempty" protobuf:"bytes,1,opt,name=scaleUpPercent"`
>
> The idea behind quantity is that all values are per minute. So if you want
> to add/remove at most 1 pod every 2 minutes you need to put 0.5 here.
>
> Quantity is used for consistency with pods. We can however use ints in
> percent-like values if you prefer.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/74525#discussion_r261774126>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVP4C-s8uVJwcXpXK-UG_wxalviTAks5vSaUCgaJpZM4bP7bw>
> .
>

mwielgus commented:
@thockin - I have worked with @gliush on this for a while. We agreed on the rough idea and the configuration parameters, so we are proceeding with the api.

This functionality has been requested multiple times already so I'm glad that we finally have someone who is working on it :).

mwielgus commented:
@thockin Then we will double the field count or create an shared denominator. Not sure what is prettier.

mwielgus commented:
@gliush Please also submit the docs as kep for this feature, as agreed.

hongshibao commented:
When is this feature planned to release?
In our internal usage of K8S, we also customize the kube-controller-manager to support similar logic with scaleDownPercent and scaleDownPods. But we put them as flags of kube-controller-manager, which means they are global settings for all HPAs.

gliush commented:
Here's a KEP (first version, need to be reviewed):
https://github.com/kubernetes/enhancements/pull/883

gliush commented:
> When is this feature planned to release?
> In our internal usage of K8S, we also customize the kube-controller-manager to support similar logic with scaleDownPercent and scaleDownPods. But we put them as flags of kube-controller-manager, which means they are global settings for all HPAs.

Well, we can combine these approaches, as in my KEP I have a "default parameters" if the User doesn't configure them explicitly.

hongshibao commented:
> > When is this feature planned to release?
> > In our internal usage of K8S, we also customize the kube-controller-manager to support similar logic with scaleDownPercent and scaleDownPods. But we put them as flags of kube-controller-manager, which means they are global settings for all HPAs.
> 
> Well, we can combine these approaches, as in my KEP I have a "default parameters" if the User doesn't configure them explicitly.

Yes. Your approach is a more general solution. Looking forward to using this new feature : )

thockin commented:
I'm not concerned with field count. and while a VERY MUCH appreciate contributors, we must be cautious with API.

I'll go read the kep, but we should come up with a convention for rational values.  I think I'd rather see something like:

```
{
    ScaleUp HPAConstraintValue
    ScaleDown HPAConstraintValue
}

type HPAConstraintValue struct {
  // One of the following must be set, but for APi compat reasons it's possible that none may be set, which is interpreted as <somethign, you tell me...>
  Pods int32
  Percent int32

  PeriodSeconds int32
}
```

so you can say:

```
constraints:
  - scaleUp:
      pods: 3
      periodSeconds: 1
  - scaleDown:
      pods: 1
      periodSeconds: 3
```

gliush commented:
@thockin: I like your idea; your configuration is more clear.
I fixed it, could you check?

gliush commented:
Just an update.
According to the discussion in the KEP https://github.com/kubernetes/enhancements/pull/883 I've updated the API.
I've added `DelaySeconds` parameter to the constraints. It makes the `StabilizationWindow` concept useless. And the command line parameter `--horizontal-pod-autoscaler-downscale-stabilization-window` from now on will specify `constraints.scaleDown.delaySeconds` default value.
Check the last commit https://github.com/kubernetes/kubernetes/pull/74525/commits/8d5a09bdde8b151c90705f0e4dbe8b7a3e84c0fd and the discussion in the KEP PR.

mathewmoon commented:
@gliush 
Good work man. This is desperately IMO. Sorry if I should be able to infer this from what is already posted, but will this be a minor version update or will it not be merged into a release until later? I am needing this functionality as soon as possible.

gliush commented:
@mathewmoon: sorry, I'm not familiar with how versioning works in k8s.
Most likely, it will be merged to the minor version release after it is fully implemented and tested.

gliush commented:
@thockin: Hi, could you approve my API change so that I could start working on the implementation?
I've received an approve for my KEP PR: https://github.com/kubernetes/enhancements/pull/883

php-coder commented:
@thockin gently ping

gliush commented:
I'm starting coding, we will discuss the result instead of discussing the API.

php-coder commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -652,15 +689,96 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
+
+// getScaleUpChangePerPeriod returns number of pods added during the constrate rate period
+func getScaleUpChangePerPeriod(scaleUpEvents map[string][]timestampedScaleEvent, args NormalizationArg) int32 {
+	cutoff := time.Now().Add(-(time.Duration)(*args.ScaleUpConstraint.Rate.PeriodSeconds))
```
> time.Now().Add(-(time.Duration)(*args.ScaleUpConstraint.Rate.PeriodSeconds))

Could we use `Sub(duration)` here instead of `Add(-duration)`?

Another option to make it easier to read and understand from the first sight would be to introduce a local variable (haven't been tested):

```go
period := time.Duration(*args.ScaleUpConstraint.Rate.PeriodSeconds)
cutoff := time.Now().Add(-period)
```

What do you think, does it worth it?

robstrong replied:
`Sub(duration)` returns a `Duration`, not a `Time`. 

You example seems more readable to me.

gliush replied:
Yep, will do.

php-coder commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -652,15 +689,96 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
+}
+
+// getScaleDownChangePerPeriod returns number of pods removed during the constrate rate period
+func getScaleDownChangePerPeriod(scaleDownEvents map[string][]timestampedScaleEvent, args NormalizationArg) int32 {
```
`getScaleUpChangePerPeriod()` and `getScaleDownChangePerPeriod()` are identical and I suggest to use a single function instead. For example, `getScaleChangePerPeriod()`.

If you want to keep 2 functions with meaningful names, consider  introducing a 3rd method with a generic logic.

gliush replied:
I keep them separate as I haven't implemented filling the "scaleUpEvents" and "scaleDownEvents" in the HorizontalController. And that might require additional changes.
So I keep it as it is for now.

gliush commented:
Hm, all the tests fail with the following error:
```
failed to convert the given HPA to autoscaling/v2beta2: failed to convert the given object back to the external version: converting (autoscaling.HorizontalPodAutoscaler).Metrics to (v2beta2.HorizontalPodAutoscaler).Metrics: Constraints not present in src
```
I'll be thankful if If anybody explains how to fix that.

php-coder commented:
@gliush The PR modifies only the external type and doesn't change the internal one. That's why conversion between these types fails. So, `pkg/apis/autoscaling/types.go` has to be updated. After that the code for conversion should be generated.

Unfortunately, I can't help with exact commands as I don't remember them and they could change over time but I'll try to run the tests and they could give you clues on what commands to execute (something like `make generate-conversion`). Also, other members will be able to look on the results of the tests to help you.

/ok-to-test

php-coder commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -703,52 +724,207 @@ func (a *HorizontalController) normalizeDesiredReplicas(hpa *autoscalingv2.Horiz
-	var possibleLimitingReason string
+// getScaleDownChangePerPeriod returns number of pods removed during the constrate rate period
+func getScaleDownChangePerPeriod(scaleDownEvents map[string][]timestampedScaleEvent, args NormalizationArg) int32 {
+	period := time.Second * time.Duration(*args.ScaleUpConstraint.Rate.PeriodSeconds)
```
Copy&paste bug has been introduced in ac5bb47c83e34cadaf86ca380a7048bb3b8d588f commit. Here is the trivial fix:
```suggestion
	period := time.Second * time.Duration(*args.ScaleDownConstraint.Rate.PeriodSeconds)
```

gliush replied:
Nice catch, thank you!
This is going to be a large PR. I haven't even started creating tests for the new functionality.

gliush commented:
I’ve got a problem with `./pkg/controller/podautoscaler/horizontal_test.go` while implementing the following KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

The test works the following way:
1. It fakes the InformerFactory to return each tests data
2. It runs the HPAController
3. It checks the result of the controller by verifying the result with the expected result from tests data
4. It stops everything.

It turned out that HPAController may run the `reconcile` function twice or even 3 times for each test. And it fails for me for the second run as I’m adding some “historic information” to implement the “periodSeconds” (check the “Proposal” section of the KEP).
An ideal solution for me would be:
- do not let HPAController run the second time
- add some latency between the HPAController cycles. So the second run won’t happen as the test will stop everything after the first run. 
But I’m not sure how to do that. 
I will appreciate any advice.

josephburnett commented:
@gliush the "historic information" is the maps you describe here, right? https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md#hpa-controller-state-changes

Generally, I would expect the HPA reconcile method to be idempotent.  Getting called multiple times should not affect the outcome of your calculations.

> it fails for me for the second run as I’m adding some “historic information” to implement the “periodSeconds” 

Can you tell me more about this?  Let's chat over Slack since we're in similar timezones and I'll report back here.

josephburnett commented:
We chatted for a bit and will do the following:

1. Find a way to make the reconcile loop run just once for simple state transition tests
2. Make sure we have test coverage of state transitions sensitive to time boundaries.  E.g. a test that verifies conditions don't change on a second reconcile loop if it happens *before* the scale limits allow further changes.
3. Make sure we have stable reconciliation test.  E.g. make sure conditions don't change when there is no need to scale, even with multiple reconciliation loops.  This is more general but important to guard against reconciliation loops.

I'll see if I can figure out a way to make the reconcile method run just once (#1).  @gliush I think it would be useful to write down what conditions you would expect to see under various states.  Perhaps an additional PR to https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md?

gliush commented:
For the history.
The problems with the tests are:
1. The HPA controller can run the reconcile function several times even for the test when it should be run once. It happens because of race condition between the [go routine that runs the hpaController and stops the controller](https://github.com/kubernetes/kubernetes/blob/9120d427dab4a69ddcf22597f81b574fb1af96bc/pkg/controller/podautoscaler/horizontal_test.go#L726) and the [routine that mocks the functions that stops the test](https://github.com/kubernetes/kubernetes/blob/9120d427dab4a69ddcf22597f81b574fb1af96bc/pkg/controller/podautoscaler/horizontal_test.go#L367)
It wasn't a problem previously as HPA controller didn't change the state before.
Now it becomes a problem, and we can't control how many times the reconcile function (and result verification check) is done.
2. To make the reconcile method idempotent we should have both the HPA controller and HPA object have the same states for all the runs. Currently HPA object remain the same, while the HPA controller state is updated between the calls (it collects the `scaleUpEvents` and `scaleDownEvents` history).


Here's an example how the number of replicas set by HPA can be different on second reconcile run.
Let’s consider the following test example:

Test data:
  initialReplicas:         5,
  maxReplicas:           20,
  metrics are set so that desired number of replicas = 15

1. First reconcile run calculates the desired number of replicas = 15. No scale history yet. The scale up limit is 10 (= currentReplicas*scaleUpLimitFactor = 5 * 2 = 10), so the controller set *desiredReplicas=10*.
And hpa controller “scaleUpEvents” is updated to contain information that we increased number of pods by 5 (newReplicas - prevReplicas = 10 - 5 = 5).

2. Second reconcile run works with the same HPA object, but different HPA controller object (that contains scaleUpEvents from previous runs). HPA controller calculates the desired number of replicas = 15. scaleUpEvents contain info about “5 pods” added since the beginning of period. During the `constraint.rate.PeriodSeconds` for `initialReplicas=5` we cannot increase number of replicas for more than 5. As we already added this number of replicas, we can’t add more. So the result of the second reconcile run is *desiredReplicas=5*

So, overall, the second run of the controller will return different result.

gliush commented:
@josephburnett : I've created a new PR to the KEP:
https://github.com/kubernetes/enhancements/pull/1108

gliush commented:
I will appreciate help with the following problem I've found in the tests:
The conversion generation warns me:
```
$ _output/bin/conversion-gen --extra-peer-dirs k8s.io/kubernetes/pkg/apis/core,k8s.io/kubernetes/pkg/apis/core/v1,k8s.io/api/core/v1 --v 1 --logtostderr -i k8s.io/kubernetes/pkg/apis/apps/v1beta1,k8s.io/kubernetes/pkg/apis/apps/v1beta2,k8s.io/kubernetes/pkg/apis/autoscaling/v1,k8s.io/kubernetes/pkg/apis/autoscaling/v2beta1,k8s.io/kubernetes/pkg/apis/autoscaling/v2beta2,k8s.io/kubernetes/pkg/apis/extensions/v1beta1 -O zz_generated.conversion
E0620 19:58:00.510116   43376 conversion.go:667] Warning: could not find nor generate a final Conversion function for k8s.io/kubernetes/pkg/apis/autoscaling.HorizontalPodAutoscalerSpec -> k8s.io/api/autoscaling/v2beta1.HorizontalPodAutoscalerSpec
E0620 19:58:00.510154   43376 conversion.go:668]   the following fields need manual conversion:
E0620 19:58:00.510157   43376 conversion.go:670]       - Constraints
```
And the file `pkg/apis/autoscaling/v2beta1/zz_generated.conversion.go` contain the following diff after that:
```
-// Convert_autoscaling_HorizontalPodAutoscalerSpec_To_v2beta1_HorizontalPodAutoscalerSpec is an autogenerated conversion function.
-func Convert_autoscaling_HorizontalPodAutoscalerSpec_To_v2beta1_HorizontalPodAutoscalerSpec(in *autoscaling.HorizontalPodAutoscalerSpec, out *v2beta1.HorizontalPodAutoscalerSpec, s conversion.Scope) error {
-       return autoConvert_autoscaling_HorizontalPodAutoscalerSpec_To_v2beta1_HorizontalPodAutoscalerSpec(in, out, s)
-}
-
```

So the code couldn't compile without it.
If I add the same function to the `pkg/apis/autoscaling/v2beta1/conversion.go`
```
+func Convert_autoscaling_HorizontalPodAutoscalerSpec_To_v2beta1_HorizontalPodAutoscalerSpec(in *autoscaling.HorizontalPodAutoscalerSpec, out *autoscalingv2beta1.HorizontalPodAutoscalerSpec, s conversion.Scope) error {
+       return autoConvert_autoscaling_HorizontalPodAutoscalerSpec_To_v2beta1_HorizontalPodAutoscalerSpec(in, out, s)
+}
```
... it doesn't help. The generated file doesn't change and the generator shows the following additional warning:

```
E0620 19:57:58.024720   43376 conversion.go:184] Rename function k8s.io/kubernetes/pkg/apis/autoscaling/v2beta1 Convert_autoscaling_HorizontalPodAutoscalerSpec_To_v2beta1_HorizontalPodAutoscalerSpec -> Convert_autoscaling_HorizontalPodAutoscalerSpec_To_unnameable_Unsupported to match expected conversion signature
```
@php-coder: do you know, what it is?  ^^

php-coder commented:
@gliush I'd try to add a conversion functions for `HPAScaleConstraints` type to conversion.go file and register them there in `scheme.AddConversionFuncs()`

When the generator will know how to convert this member, I hope that the conversion for `HorizontalPodAutoscalerSpec` will be generated automatically.

Hope this helps, otherwise let me know.

gliush commented:
I'm mostly done and I encourage everybody to have a look at my changes!
I've reworked all the commits, now I have only 5 of them to simplify reading.
I'm going to make a full reread later this week to be sure that everything is correct.

I'm not sure that I should introduce a separate module for the newest HPA implementation as was discussed [in the email thread](https://groups.google.com/forum/#!topic/kubernetes-sig-architecture/pjtpgefOC2M). If you provide good arguments for that I can rework everything.

I had to add plenty of unit tests to cover all code paths. Though, I still want to rework the testing infrastructure as described in https://github.com/kubernetes/kubernetes/issues/79222

I can't cover all the status behavior changes because it requires a full testing infrastructure rewriting (see the issue link above).

I've got one question.
I see that HPA has a "MetricsStatuses" to show how "Metrics" have been used.
I don't think that we need to have "ConstraintStatuses" by analogy. If anybody provides arguments about why it might be useful, I will add them.

gliush commented:
@mwielgus @josephburnett: Could you have a look?

gliush commented:
> Hope this helps, otherwise let me know.

@php-coder: Thanks for your advice, it helped a lot!!

josephburnett commented:
/assign

As mentioned in the SIG Autoscaling meeting, I'll review this PR this week.

josephburnett commented:
Some initial high-level comments:
1. The API should take into account some of @thockin's [comments](https://github.com/kubernetes/enhancements/pull/883#pullrequestreview-232695825), in particular the suggestion to use more generic policy entries: https://github.com/kubernetes/enhancements/pull/883#discussion_r280170553
2. There are a lot of changes to the `normalizeDesiredReplicas` and `stabilizeRecommendation` methods. I recommend forking those methods into `normalizeDesiredReplicasAlpha` and `stabilizeRecommendationAlpha` to make the changes. And to use the alpha methods only when the new constraints (or policies) fields are present. This is based on a conversation with SIG Architecture: https://groups.google.com/forum/#!topic/kubernetes-sig-architecture/pjtpgefOC2M

josephburnett commented:
I've chatted with @gliush about how to safely separate the new horizontal controller code.  We came to the conclusion that we should fork the `reconcileAutoscaler` method and everything below (in the call graph) into two separate files.  One file would contain the current, stable code.  The other would containe the new, restructured code that supports constraints

For reference, the call graph is roughly:
```
processNextWorkItem()
reconcileKey(key)
reconcileAutoscaler(hpav1Shared, key)     <=== choose which version based on presence of new fields
computeReplicasForMetrics(hpa, scale, metricSpecs)
computeReplicasForMetric(hpa, spec, specReplicas, statusReplicas, selector, status)
computeStatusForResourceMetric(currentReplicas, metricSpec, hpa, selector, status)
replicaCalc.GetResourceReplicas(currentReplicas, targetUtilization, name, namespace, selector)
```

I think that `reconcileAutoscaler` is a good cut point because recent that's where the complexity starts to increase dramatically.  Recent bugs were entirely in and below that method (https://github.com/kubernetes/kubernetes/pull/79035) and new features such as scale-to-zero are also at and below that method (https://github.com/kubernetes/kubernetes/pull/74526).

Once the new fields prove to be stable, the old code would be deleted and the new code would become the default path.

arjunrn commented:
@gliush Thanks for the PR. I see that you have made changes to the stabilization section where even scale up is stabilized and it is stabilized in the interval of `max(constraintUpWindow, constraintDownWindow)`. This change effectively means that when the constraints are set there will be no scale up or scale down after a scaling action is performed for the max time of the 2 windows. This is problem when you want to scale up/down quickly but slowly in the opposite direction. Also in the original code the stabilization is only applied to scaling down and I think we should not change that because that cover 99% of the use cases.
Secondly, the scaling events are stored in memory only. Which means when the controller is restarted(which could happen in response to cluster autoscaling or node upgrades) all the data is lost and the guarantees are lost. Also storing them entirely in memory makes it harder to trace and debug the actions performed by the autoscaler. I would recommend persisting them somehow. Either by have something like a `HorizontalPodAutoscalerHistory` object or storing them in the status(could be a problem when large periods are specified).

shaoxt commented:
The feature looks useful to me. In our case, we need to make the scale up speed faster.
These settings will help.

tedyu commented on pkg/apis/autoscaling/types.go:
```diff
@@ -92,11 +92,73 @@ type HorizontalPodAutoscalerSpec struct {
+	//			for scaleDown: Pods = -1
+	// more info could be found in the KEP https://github.com/kubernetes/enhancements/pull/883
+	// +optional
+	Pods *int32
```
I wonder why the fields are pointers .
A nil pointer can be represented by -1 if primitive type is used.

mwielgus commented:
Just FYI. Code Freeze for 1.16 is coming this Thursday.

RainbowMango commented on pkg/apis/autoscaling/types.go:
```diff
@@ -92,11 +92,73 @@ type HorizontalPodAutoscalerSpec struct {
 	// more information about how each type of metric must respond.
 	// +optional
 	Metrics []MetricSpec
+	// constraints contain HPA object configuration parameters
```
```suggestion
	// Constraints contains HPA object configuration parameters
```
a nit

vllry commented:
/cc

gliush commented:
Sorry, I was very busy, now I'm back.
I'm going to continue working on this ticket:
- check the [KEP modification](https://github.com/kubernetes/enhancements/pull/1234/files) by @arjunrn 
- check the [PR rebase](https://github.com/kubernetes/kubernetes/pull/82256) by @arjunrn
- will coordinate with @arjunrn to complete the work.

josephburnett commented on pkg/apis/autoscaling/types.go:
```diff
@@ -95,6 +95,66 @@ type HorizontalPodAutoscalerSpec struct {
+	// The scaling behavior while scaling down. Contains a list of policies and the policy selector.
+	// +optional
+	ScaleDown *HPAScalingDirectionBehavior
+}
```
The names here are bit different from the KEP: https://github.com/kubernetes/enhancements/pull/1234/files#diff-fd841be6aa8e9bedba8f79659f684798R455  Is that by design?

gliush replied:
Oh, I missed it! Thank you!
Will fix it.

gliush replied:
Fixed.

gliush committed:
Introduces all API changes needed for Configurable HPA PR

gliush committed:
Generates boilerplate code

gliush committed:
Adds the algorithm implementation for the Configurable HPA

josephburnett commented on pkg/apis/autoscaling/types.go:
```diff
@@ -95,6 +95,68 @@ type HorizontalPodAutoscalerSpec struct {
+	StabilizationWindowSeconds *int32
+	// SelectPolicy is used to specify which policy should be selected after evaluation.
+	// +optional
+	SelectPolicy *ScalingPolicySelect
```
I'm glad you made this a dedicated type instead of a `*string` as in the KEP.  
![image](https://user-images.githubusercontent.com/1103629/66808968-ace81080-ef2c-11e9-8be0-1124e7a603af.png)
You might consider updating the KEP to match.

gliush replied:
We discussed that and decided to merge the KEP as it was because KEP is just a proposal, not the documentation/API change itself.
If you feel like we should update the KEP, we can definitely do it.

josephburnett replied:
After everything is done, it would be nice.  But not necessary.  :)

gliush replied:
We'll update the KEP as soon as the API is stabilized! :)

josephburnett commented on pkg/apis/autoscaling/types.go:
```diff
@@ -95,6 +95,68 @@ type HorizontalPodAutoscalerSpec struct {
+	// Value contains the amount of change which is permitted by the policy.
+	Value *int32
+	// PeriodSeconds specifies the window of time for which the policy should hold true.
+	PeriodSeconds *int32
```
I would prefer to make these `int32` (no pointer) as in the KEP.  This indicates they are required.  An incomplete `HPAScalingPolicy` should be invalid (defaults at this level would be confusing).

![image](https://user-images.githubusercontent.com/1103629/66809105-fd5f6e00-ef2c-11e9-9003-8aa2f75f843a.png)

josephburnett replied:
@gliush @arjunrn and I discussed this in Slack.  We concluded that `Value` and `PeriodSeconds` should indeed be `int32` (no pointer).  Our intent is that a user must provide all parts of the policy explicitly (no defaults) to make the API less confusing.  So we will validate that `Value` and `PeriodSeconds` are **not** `0`.  And that `Type` is not `""`.

The open question was whether a `Value` of `0` is a valid use case.  It's included in the KEP as one: https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md#story-4-scale-up-as-usual-do-not-scale-down.  We think it could be, so the API should allow for it.

However we prefer not to use pointers.  They are cumbersome in golang because they require an intermediate variable.  They add a layer of indirection and risk for nil dereference.  And they make the API less intuitive (e.g. what does `nil` mean?).  We decided that a `SelectPolicy` value of `"disabled"` would be a more intuitive way of turning off up or down scaling.  And would allow us to avoid a `*int32` for `Value`.

For `PeriodSeconds`, `0` simply doesn't make sense (divide by 0).  So it doesn't need a pointer either.

gliush replied:
Just one note here. For "Type" we check that it has an allowed value ("max", "min" or "disabled"), and not a simple `is not ""`

josephburnett commented on pkg/apis/autoscaling/types.go:
```diff
@@ -95,6 +95,68 @@ type HorizontalPodAutoscalerSpec struct {
+	// and scaling down. It also is used to customize other aspects of scaling like the
+	// scale down stabilization window
+	// +option
+	Behavior *HorizontalPodAutoscalerBehavior
```
Who uses this `pkg/apis/autoscaling/types.go` file?  All of the autoscaling imports I could find in the Kubernetes repo reference a versioned type.  E.g. `pkg/apis/autoscaling/v2beta2`.  Why the duplication?

gliush replied:
I'm not sure. We might need to add a ticket to investigate this and remove the duplication if possible.

tedyu commented on pkg/apis/autoscaling/v1/conversion.go:
```diff
@@ -383,6 +391,15 @@ func Convert_v1_HorizontalPodAutoscaler_To_autoscaling_HorizontalPodAutoscaler(i
 		delete(out.Annotations, autoscaling.MetricSpecsAnnotation)
 	}
 
+	if behaviorEnc, hasConstraints := out.Annotations[autoscaling.BehaviorSpecsAnnotation]; hasConstraints {
```
out.Annotations[autoscaling.BehaviorSpecsAnnotation] may have been assigned on line 355.
In that case, is json.Unmarshal call below necessary ?

gliush replied:
@tedyu: Probably I didn't get your question, could you rephrase it?

line 355 is part of `Convert_autoscaling_HorizontalPodAutoscaler_To_v1_HorizontalPodAutoscaler` function. I.e., it converts the `autoscaling.HorizontalPodAutoscaler` object of the current version (with behavior) to the v1 version. The v1 version doesn't have behaviors (it's a base version), so that's why we put `behavior` into annotations.

The current line (394), is part of `Convert_v1_HorizontalPodAutoscaler_To_autoscaling_HorizontalPodAutoscaler`, i.e. it converts the object back from v1 to the current version. And if there's information about behavior in annotation, we convert it into a usual `autoscaling.HorizontalPodAutoscalerBehavior` field (check line 399).

tedyu replied:
Due to the way diff is showed, I didn't pay attention to the two func's (as you described).

Thanks for explaining.

arjunrn commented:
@thockin We have made significant changes to the API and now have a list of policies as you previously suggested. Can you have another look? The original [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md) has also been updated with the newer changes.

arjunrn commented:
/label api-review

josephburnett commented:
This API change is already labeled with `api-review` and is in the [project board](https://github.com/orgs/kubernetes/projects/13) assigned to @thockin.

josephburnett commented:
/lgtm
/approve
... from the autoscaling side.

Nice work @gliush and @arjunrn!  There was a lot of thought, discussion and hard work that went into this API change.

thockin commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -117,6 +120,64 @@ type MetricSpec struct {
+
+// HPAScalingRules configures the scaling policy and the policy selector
+type HPAScalingRules struct {
+	// StabilizationWindowSeconds is the number of seconds for which past recommendations should be considered while
```
Document what the behavior is when not specified (default to 0?)

This is true for EVERY optional field.

thockin replied:
I don't see any code here setting defaults, so it really needs to be defined

gliush replied:
Fixed: https://github.com/kubernetes/kubernetes/pull/74525/commits/877d2ffeccb6d8fd1e47a12eab02de44b8bac07b

thockin commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -117,6 +120,64 @@ type MetricSpec struct {
+	// scaling up or scaling down
+	// +optional
+	StabilizationWindowSeconds *int32 `json:"stabilizationWindowSeconds" protobuf:"varint,3,opt,name=stabilizationWindowSeconds"`
+	// SelectPolicy is used to specify which policy should be used
```
Default when not specified

gliush replied:
Fixed: https://github.com/kubernetes/kubernetes/pull/74525/commits/877d2ffeccb6d8fd1e47a12eab02de44b8bac07b

thockin commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -117,6 +120,64 @@ type MetricSpec struct {
+
+const (
+	// PodsScalingPolicy is a policy used to specify a change in absolute number of pods.
+	PodsScalingPolicy HPAScalingPolicyType = "pods"
```
Constants start with caps.

thockin replied:
Is "Pods" the right name?  HPA already says "pods".  Does "Count" make more sense?

gliush replied:
We want to have two different policy types: for the number of pods and for percentage.
I suppose, "Pods" suites better, as the json representation will be more self-descriptive:
`policies: [{'type': 'Pods', 'value': 10, 'periodSeconds': 60}]`

thockin commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -117,6 +120,64 @@ type MetricSpec struct {
+type HPAScalingPolicy struct {
+	// Type is used to specify the scaling policy.
+	Type HPAScalingPolicyType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=HPAScalingPolicyType"`
+	// Value contains the amount of change which is permitted by the policy.
```
Is there any validation for this? Is 0 allowed? Are negative numbers allowed? Is there an upper bound? What if I specify 100000000 ?

gliush replied:
Negative numbers are not allowed, validation is part of another PR (the current one is about API change):
https://github.com/gliush/kubernetes/pull/2/files#diff-4364e3f754b12e39f60eea33105f6c7eR203
it is int32, any positive number is correct. It does not have a lot of sense, though, I do not see a reason why we should forbid that.

gliush replied:
I've added validation to the PR: https://github.com/kubernetes/kubernetes/pull/74525/commits/c32de0c8f34fc64fdec96dd84e31c2c01ce47f1f

thockin commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -117,6 +120,64 @@ type MetricSpec struct {
+	Type HPAScalingPolicyType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=HPAScalingPolicyType"`
+	// Value contains the amount of change which is permitted by the policy.
+	Value int32 `json:"value" protobuf:"varint,2,opt,name=value"`
+	// PeriodSeconds specifies the window of time for which the policy should hold true.
```
Is there any validation for this? Is 0 allowed? Are negative numbers allowed? Is there an upper bound? What if I specify 100000000 ?

gliush replied:
We are going to add validation to limit the periodSeconds with some reasonable amount of time (60min?). It is being done.

arjunrn replied:
@thockin we want to have a a flag in the controller so that cluster operators can specify the maximum allowed value(because this affects how many events are stored in the controller, by default we want to restrict it to 30 mins). Since we don't have access to the controller values in the validation at this stage we can only validate that it is a positive value. If a value which is too large is specified an HPA will be created but it will stay disable/inactive.

gliush replied:
I've added validation that will restrict periodSeconds to 30min

thockin commented on pkg/apis/autoscaling/v2beta1/conversion.go:
```diff
@@ -289,3 +294,46 @@ func Convert_v2beta1_PodsMetricStatus_To_autoscaling_PodsMetricStatus(in *autosc
+		}
+	}
+
+	if in.Spec.Behavior != nil {
```
These 2 blocks are both testing for nil - why 2 blocks?

gliush replied:
Fixed: https://github.com/kubernetes/kubernetes/pull/74525/commits/877d2ffeccb6d8fd1e47a12eab02de44b8bac07b

thockin commented on pkg/apis/autoscaling/v2beta1/conversion.go:
```diff
@@ -289,3 +294,46 @@ func Convert_v2beta1_PodsMetricStatus_To_autoscaling_PodsMetricStatus(in *autosc
+		return err
+	}
+	if in.Spec.Behavior != nil {
+		old := out.Annotations
```
Why this copy?

gliush replied:
I copied the implementation from here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/autoscaling/v1/conversion.go#L327

Ok, I'll not use it here.

gliush replied:
Sorry, I realized that I don't understand the question. Could you rephrase, please?

arjunrn replied:
I think @thockin is referring to the statement `old := out.Annotations`. It actually does not copy the map but creates another pointer. Overall this `if` block seems superfluous and can be dropped.

gliush replied:
Fixed: https://github.com/kubernetes/kubernetes/pull/74525/commits/877d2ffeccb6d8fd1e47a12eab02de44b8bac07b

thockin commented on pkg/apis/autoscaling/v2beta1/conversion.go:
```diff
@@ -289,3 +294,46 @@ func Convert_v2beta1_PodsMetricStatus_To_autoscaling_PodsMetricStatus(in *autosc
+		if err != nil {
+			return err
+		}
+		out.Annotations[autoscaling.BehaviorSpecsAnnotation] = string(behaviorEnc)
```
What if this annotation already existed?  This logic needs to handle every pathology you can think up.  It may be aceptable to nuke it (assume ownership), but please document that here.

gliush replied:
I'm not an expert in this "conversion" mechanics, but I doubt that this might cause any problem. I'd like to ask for a piece of advice here from an expert.
1. Is it possible that this annotation exists? This is a conversion from the most recent "HPA" spec to the v2beta1 spec. The most recent spec contains behavior, and we don't have any code to generate the annotation with behavior in it.
2. Even if the annotation for behavior exists, we will just rewrite it and increase the size of the "out.Annotation"  map.
3. I repeated the "Annotation" generation from here: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/autoscaling/v1/conversion.go#L327

gliush replied:
Added a comment about possible overwriting the annotation.

arjunrn replied:
Can you also mention it in the __User facing changes__ section of the PR description.

thockin reviewed (commented):
Overall the comments-as-docs are very under-specified.  Please run through all of those and make sure that a new user who was exploring this API for the first time, using these docs, can figure out what happens.

Any optional field needs to say what happens if not specified.

Any list field needs to define what happens when empty.

Any list fields needs to define the upper-bound for number of members.

Any scalar field needs to define what values are valid (e.g negative ints, zero values, empty strings).

I don't see defaulting code.  I don't see validation code.  Assume a pathological user who is going to send all sorts of garbage into this API and try to crash the controllers or nodes.  Defend.

arjunrn commented on pkg/apis/autoscaling/types.go:
```diff
@@ -96,20 +96,31 @@ type HorizontalPodAutoscalerSpec struct {
-	// and scaling down. It also is used to customize other aspects of scaling like the
-	// scale down stabilization window
-	// +option
+	// Behaviour contains the scaling behavior for the HPA
```
The comment for the behavior field should contain information about what the behavior field is a wrapper for. In this case it is the top level field to contain all the values which customize how the controller scales the target. Specific are not necessary at this stage because they can be explained in the comments of individual fields. Similar the defaults here need not be specified in full detail because each of the fields has it's own defaults.

gliush replied:
@thockin explicitly asked me to add information about default values for any non-specified field. And I'm not sure how to do that properly:
- If I don't specify the default `behavior` here, it might confuse the reader. 
- If I specify it here, it means, that I should explain all underlying abstractions at this point (`HPAScalingRules`, `HPAScalingPolicy`, etc)
- Moreover, in any case, I will have several places where I put the same "default values" (i.e. [behavior usage](https://github.com/kubernetes/kubernetes/pull/74525/files?file-filters%5B%5D=.go&file-filters%5B%5D=.json&file-filters%5B%5D=.pb&file-filters%5B%5D=.proto&file-filters%5B%5D=.yaml#diff-e9b775adb394bd5cd6c7606650443d57R102) and [behavior declaration](https://github.com/kubernetes/kubernetes/pull/74525/files?file-filters%5B%5D=.go&file-filters%5B%5D=.json&file-filters%5B%5D=.pb&file-filters%5B%5D=.proto&file-filters%5B%5D=.yaml#diff-e9b775adb394bd5cd6c7606650443d57R116) )
@thockin : could you give an advice?

gliush replied:
I've updated the comment

arjunrn commented on pkg/apis/autoscaling/types.go:
```diff
@@ -96,20 +96,31 @@ type HorizontalPodAutoscalerSpec struct {
 
 // HorizontalPodAutoscalerBehavior configures a scaling behavior for Up and Down direction
-// (scaleUp and scaleDown fields respectively)
+// (scaleUp and scaleDown fields respectively). Check HPAScalingRules comment for more info.
```
Generally comments should not refer to other comments are should be self-contained. In this case I would said that the comment should say something like _configures the scaling behavior of the target in both up and down directions_.

gliush replied:
I've changed that already.

arjunrn commented on pkg/apis/autoscaling/types.go:
```diff
@@ -96,20 +96,31 @@ type HorizontalPodAutoscalerSpec struct {
-	// The scaling behavior while scaling up. Contains a list of policies and the policy selector.
+	// scaling policy for scaling Up
+	// if not set, the default value is used:
+	//   {selectPolicy: "Max", stabilizationWindowSeconds=0, policies: [{"Pods", 4, 60sec}, {"Percent", 100, 60sec}]
```
Explain the defaults in plain text only like [here](https://github.com/kubernetes/kubernetes/blob/877d2ffeccb6d8fd1e47a12eab02de44b8bac07b/pkg/apis/apps/types.go#L416)

gliush replied:
Done.

arjunrn commented on pkg/apis/autoscaling/types.go:
```diff
@@ -119,17 +130,25 @@ type ScalingPolicySelect string
 
-// HPAScalingRules configures the scaling policy and the policy selector
+// HPAScalingRules configures the scaling behavior for one direction.
+// This Rules are applied after calculating DesiredReplicas from metrics for the HPA.
```
`DesiredReplicas` is an internal detail of the controller and I don't think it should be mentioned here. If absolutely necessary then explain that these rules restrict the change in the number of replicas as the target is scaled towards its desired size.

gliush replied:
It is not. Well, partly.
At least, it is set in the status, so users know about it.

gliush replied:
https://github.com/kubernetes/kubernetes/blob/c32de0c8f34fc64fdec96dd84e31c2c01ce47f1f/pkg/apis/autoscaling/types.go#L339

arjunrn commented on pkg/apis/autoscaling/types.go:
```diff
@@ -119,17 +130,25 @@ type ScalingPolicySelect string
+// This Rules are applied after calculating DesiredReplicas from metrics for the HPA.
+// They can limit the scaling velocity by specifying scaling policies.
+// They can prevent flapping by specifying the stabilization window, so that the
+// number of replicas is not set instantly, instead, the safest value from the stabilization
```
reword -> `so that the number of replicas is not changed instantly when the metrics change but the safest value..`

gliush commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -117,6 +120,64 @@ type MetricSpec struct {
+
+const (
+	// MaxPolicySelect selects the policy with the highest possible change
+	MaxPolicySelect ScalingPolicySelect = "max"
```
Fixed: https://github.com/kubernetes/kubernetes/pull/74525/commits/877d2ffeccb6d8fd1e47a12eab02de44b8bac07b

gliush commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -117,6 +120,64 @@ type MetricSpec struct {
+// HorizontalPodAutoscalerBehavior configures a scaling behavior for Up and Down direction
+// (scaleUp and scaleDown fields respectively)
+type HorizontalPodAutoscalerBehavior struct {
+	// constraint value for scaling Up
```
I've added more explanation here, could you check it: https://github.com/kubernetes/kubernetes/pull/74525/commits/877d2ffeccb6d8fd1e47a12eab02de44b8bac07b

I've tried my best making it clear and concise, could you share your ideas on how to improve it?

gliush commented on staging/src/k8s.io/api/autoscaling/v2beta2/types.go:
```diff
@@ -72,6 +72,9 @@ type HorizontalPodAutoscalerSpec struct {
 	// If not set, the default metric will be set to 80% average CPU utilization.
 	// +optional
 	Metrics []MetricSpec `json:"metrics,omitempty" protobuf:"bytes,4,rep,name=metrics"`
+	// Behaviour contains the scaling behavior for the HPA
```
Added explanation: https://github.com/kubernetes/kubernetes/pull/74525/commits/877d2ffeccb6d8fd1e47a12eab02de44b8bac07b

gliush commented:
@thockin : I didn't add validation code to the PR, it is kept in the separate "code-change-PR". 
https://github.com/gliush/kubernetes/pull/2

If you wish, I can move the "validation" change here, into this PR.
Should I?

gliush commented:
@thockin : I've moved validation into the current PR. It really helps to understand the changes.
Could you have another look at the changes?

gliush commented:
/test pull-kubernetes-e2e-gce-100-performance

arjunrn commented on pkg/apis/autoscaling/v2beta1/conversion.go:
```diff
@@ -257,3 +259,45 @@ func Convert_v2beta1_PodsMetricStatus_To_autoscaling_PodsMetricStatus(in *autosc
+		behaviorEnc, err := json.Marshal(in.Spec.Behavior)
+		if err != nil {
+			return err
+		}
```
This entire block of code is redundant. There are other places where similar code is used but it really adds nothing.

gliush replied:
@arjunrn : what block do you mean? lines 275-277?

gliush commented:
@thockin : Ping. Could you have a look?!
I want this PR to be part of 1.17, we're almost out of the time.

thockin commented on pkg/apis/autoscaling/types.go:
```diff
@@ -95,6 +95,90 @@ type HorizontalPodAutoscalerSpec struct {
+type HorizontalPodAutoscalerBehavior struct {
+	// scaleUp is scaling policy for scaling Up.
+	// If not set, the default value is used:
+	// - The first policy is to increase no more than 4 pods per 60 seconds.
```
I would rephrase as something like:

```
// If not set, the default value is the higher of:
//   * increase no more than 4 pods per 60 seconds
//   * double the number of pods per 60 seconds
// No stabilization is used.

gliush replied:
Fixed, thank you for wording.

thockin commented on pkg/apis/autoscaling/types.go:
```diff
@@ -95,6 +95,90 @@ type HorizontalPodAutoscalerSpec struct {
+	// +optional
+	ScaleUp *HPAScalingRules
+	// scaleDown is scaling policy for scaling Down.
+	// if not set, the default value is used:
```
// If not set, the default value is to allow to scale down to 0 pods, with a
// 300 second stabilization window (i.e., the highest recommendation for
// the last 300sec is used).

gliush replied:
Done.

thockin commented on pkg/apis/autoscaling/types.go:
```diff
@@ -95,6 +95,90 @@ type HorizontalPodAutoscalerSpec struct {
+	ScaleUp *HPAScalingRules
+	// scaleDown is scaling policy for scaling Down.
+	// if not set, the default value is used:
+	// - The only policy specified is to allow to scale down to 0 pods.
```
Really?  We scale to 0 by default?

gliush replied:
Well, we don't limit it somehow.
But you are right, I had an incorrect comment, we allow scaling to minReplicas, not to 0.

gliush replied:
Changed the comment.

thockin commented on pkg/apis/autoscaling/types.go:
```diff
@@ -95,6 +95,90 @@ type HorizontalPodAutoscalerSpec struct {
+)
+
+// HPAScalingRules configures the scaling behavior for one direction.
+// This Rules are applied after calculating DesiredReplicas from metrics for the HPA.
```
"These rules"

gliush replied:
Fixed.

thockin commented on pkg/apis/autoscaling/types.go:
```diff
@@ -95,6 +95,90 @@ type HorizontalPodAutoscalerSpec struct {
+type HPAScalingPolicy struct {
+	// Type is used to specify the scaling policy.
+	Type HPAScalingPolicyType
+	// Value contains the amount of change which is permitted by the policy.
```
What are the bounds on this?

gliush replied:
It must be greater than zero.
We don't have upper bound limitation for it as it is a pointless limitation that doesn't have any real reason.

gliush replied:
Added a comment about "greater than zero".

thockin commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -156,6 +159,65 @@ func validateMetrics(metrics []autoscaling.MetricSpec, fldPath *field.Path, minR
+			*rules.SelectPolicy != autoscaling.MaxPolicySelect &&
+			*rules.SelectPolicy != autoscaling.MinPolicySelect &&
+			*rules.SelectPolicy != autoscaling.DisabledPolicySelect {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("selectPolicy"), rules.SelectPolicy, "must be max, min, or disabled"))
```
`field.NotSupported` is what you want here, and you can pass it a slice of valid values.

thockin commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -156,6 +159,65 @@ func validateMetrics(metrics []autoscaling.MetricSpec, fldPath *field.Path, minR
+func validateScalingPolicy(policy autoscaling.HPAScalingPolicy, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if policy.Type != autoscaling.PodsScalingPolicy && policy.Type != autoscaling.PercentScalingPolicy {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("type"), policy.Type, "must be either pods or percent"))
```
field.NotSupported

thockin commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -156,6 +159,65 @@ func validateMetrics(metrics []autoscaling.MetricSpec, fldPath *field.Path, minR
+	if policy.PeriodSeconds <= 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("periodSeconds"), policy.PeriodSeconds, "must be greater than zero"))
+	}
+	if policy.PeriodSeconds > 1800 {
```
1800 should be a const somewhere

gliush replied:
Created a `const` with comment and used it.

thockin commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -156,6 +159,65 @@ func validateMetrics(metrics []autoscaling.MetricSpec, fldPath *field.Path, minR
+		if rules.StabilizationWindowSeconds != nil && *rules.StabilizationWindowSeconds < 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("stabilizationWindowSeconds"), rules.StabilizationWindowSeconds, "must be greater than or equal to zero"))
+		}
+		if rules.StabilizationWindowSeconds != nil && *rules.StabilizationWindowSeconds > 3600 {
```
3600 should be a const somewhere

gliush replied:
Done.

thockin reviewed (commented):
This looks MUCH better.  I still don't see any defaulting code.  In general you don't want to assume defaults in the implementation.  You can catch unset values and actually fill in the default values.  This allows different defaults for different API versions.  See pkg/apis/autoscaling/v2beta2/defaults.go

Does this need any feature gate?

Where is the implementation?  We don't want to introduce API without an impl, lest we have to revert it when the impl doesn't make  it in time.

josephburnett commented:
> Where is the implementation? We don't want to introduce API without an impl, lest we have to revert it when the impl doesn't make it in time.

FYI: I asked @gliush and @arjunrn to separate the API from the implementation to make the API changes easier to understand and review.

gliush commented:
@thockin : I've addressed all your comments.
I also added implementation from another PR (it was in gliush/kubernetes repo).

josephburnett commented:
/retest

josephburnett commented on pkg/controller/podautoscaler/horizontal_test.go:
```diff
@@ -3126,4 +4215,4 @@ func TestNoScaleDownOneMetricEmpty(t *testing.T) {
 }
 
-// TODO: add more tests
+// TODO: add more tests for stabilization
```
Hmm ... I don't really find this so useful.  If you know of missing test cases, you could leave a commented list in the table test.

gliush replied:
Oh, this is an obsolete comment. I'm removing it.

josephburnett commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -52,15 +52,56 @@ import (
+				Value:         scaleDownLimitPercent,
+				PeriodSeconds: scaleDownPeriod,
+			},
+		},
```
What existing limit is this policy replacing?  What effect will this have?

I know that the two scale up policies are replacing existing constants (limit 2x or 4 pods, whichever is larger) and the scale down stabilization window is 300 seconds.  But I'm not aware of an existing scale down limit of 100% within 60 seconds which requires this default scale down policy.

gliush replied:
This is how it works:
1. If the user doesn't specify any policy, we use the default one (that is listed here).
2. It is needed because if no policy is specified, the HPA will not scale down as it doesn't have any policy to follow:
https://github.com/kubernetes/kubernetes/pull/74525/files/0328cec975e6438009d2b9a069a034a63f9df1b7#diff-ff4e8f1a0c24ee394a0ddf37d64b2960R1098

So, the default policy allows scaling down to 0%.

josephburnett reviewed (approved):
/lgtm
/approve

thockin commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -165,6 +175,73 @@ func validateMetrics(metrics []autoscaling.MetricSpec, fldPath *field.Path, minR
+				fmt.Sprintf("must be less than or equal to %v", MaxStabilizationWindowSeconds)))
+		}
+		if rules.SelectPolicy != nil &&
+			*rules.SelectPolicy != autoscaling.MaxPolicySelect &&
```
This can be simplified to `validSelectPolicyTypes.Has(*rules.SelectPolicy)` I think

thockin replied:
Acceptable as a followup PR if there's nothing else to block this

thockin commented on pkg/apis/autoscaling/types.go:
```diff
@@ -95,6 +95,90 @@ type HorizontalPodAutoscalerSpec struct {
+// (scaleUp and scaleDown fields respectively).
+type HorizontalPodAutoscalerBehavior struct {
+	// scaleUp is scaling policy for scaling Up.
+	// If not set, the default value is the higher of:
```
OK.  My only real remaining concern is here.

You are documenting the default behavior as part of the API, but not crystallizing that into the resource itself.

What we normally do is add code to defaults.go, looking for unset values and setting them to the defaults.  This means that a user who did not specify `scaleUp` when they created the object can read the object back and see what the active values are.  It also means that you can have API v3 with a different default if you need to.

This PR is not doing this.  Instead you are assuming defaults deep inside the controller.  If you go this route you can never ever change them because the stored objects have no value set, so any change would be breaking.

I strongly urge you to do it the normal way.  Do you have good reasons to do it differently?

gliush replied:
I have implemented it as a part of defaults.go
Could you have another look?

thockin reviewed (commented):
Only 1 meaningful thing.  I'm willing to approve the API changes here if we reach agreement.  I.e.

1) You convince me that being different is correct
2) You promise to change the defaulting in a followup PR within O(days)

gliush commented:
@thockin : I am going to add the defaulting right now and fix all the tests. This will be ready today.

josephburnett commented:
/test pull-kubernetes-node-e2e-containerd

josephburnett commented:
/test pull-kubernetes-node-e2e-containerd
/test pull-kubernetes-kubemark-e2e-gce-big

gliush commented:
@thockin: It is ready, could you approve it?
What should I do to make it part of 1.17? If I understand correctly we should merge it?

josephburnett commented:
/lgtm
/approve

josephburnett commented:
/assign @thockin

josephburnett commented:
/hold

I dug into the failed pull-kubernetes-kubemark-e2e-gce-big test and saw this: 

> TestMetrics error: [restart counts violation: RestartCount(kube-controller-manager-e2e-74525-ac87c-kubemark-master, kube-controller-manager)=1, want <= 0]

I deployed these changes with `kube-up.sh` and when I add `behavior.scaleDown.stabilizationWindowSeconds: 30` and [apply load](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/) then controller-manager crash loops.  Nothing in the logs so I don't know why, but this must be fixed and requires an e2e test.

josephburnett commented:
Maybe related: https://github.com/kubernetes/kubernetes/issues/84990

thockin commented on pkg/apis/autoscaling/v2beta2/defaults.go:
```diff
@@ -48,4 +88,42 @@ func SetDefaults_HorizontalPodAutoscaler(obj *autoscalingv2beta2.HorizontalPodAu
 	}
+
+	// if behavior is specified, we should fill all the 'nil' values with the default ones
+	if obj.Spec.Behavior != nil {
```
The API spec says that `Behavior == nil` will also set the defaults.  I see the test cases don't cover that, but I think you want to create a value for behavior and then merge the defaults in, right?

thockin reviewed (commented):
I'll approve this as-is, but you really need that one extra case in defaulting - please followup.

I see there's a hold for another reason, so this may be in vain, but good luck. :)


/lgtm
/approve

josephburnett commented on pkg/apis/autoscaling/v2beta2/defaults.go:
```diff
@@ -18,11 +18,51 @@ package v2beta2
+var (
+	scaleUpLimitPercent         int32 = 100
+	scaleUpLimitMinimumPods     int32 = 4
+	scaleUpPeriod               int32 = 60
```
This should really be 15 to keep parity with existing defaults, which resync HPAs every 15 seconds.

gliush commented:
I've changed the default period seconds for HPA: 60s -> 15s to correspond to the previous behavior.
All the tests are passed.

josephburnett commented:
@arjunrn thanks for fixing the controller-manager crash!  I've done some manual testing in a cluster and it works nicely.  Added and removed behaviors which are respected.

Just two requests:
1. Please squash this PR before I give an LGTM. There are 21 commits and this repo doesn't squash for you!
2. Please provide a pointer to the E2E test for this.  @arjunrn I think you have one already.  But I want to be sure we have something that will guard against a bug like the controller-manager crashing.

gliush committed:
Adds validation rules and proper defaults

arjunrn committed:
Adds tests

Signed-off-by: Arjun Naik <arjun@arjunnaik.in>

josephburnett commented:
/lgtm

I checked and there is no difference between the latest squash and the unsquashed changes I tested.  Nice work!

josephburnett commented:
/hold cancel

arjunrn commented:
/test pull-kubernetes-kubemark-e2e-gce-big

saschagrunert commented:
Hey @gliush 👋, Im currently looking at the latest [generated releases notes](https://github.com/evillgenius75/sig-release/blob/0cc8b17805dc198561054f0e1498a122320a205f/releases/release-1.18/release-notes-draft.md) and saw that this note needs to either state the user facing change or set it to `NONE`. Do you think you can change the PR description for that? The next re-generation of the notes will fix that issue on our side later on. :)

gliush commented:
@saschagrunert : I've updated the release notes, could you check, please? I'm a little bit concerned about the style and the amount of details.

saschagrunert commented:
Thank you @gliush :pray: I assume the API is new and we should state that to the user as well, like:

```
Added the HPA API, which allows scale ...
```

So in general it is looking good, but we have include information like mentioned here: https://github.com/saschagrunert/community/blob/master/contributors/guide/release-notes.md#contents-of-a-release-note

gliush commented:
@saschagrunert : thank you! Done!

saschagrunert commented:
Wonderful, thank you again! :)
